### PR TITLE
Mssw chapter10

### PR DESCRIPTION
**Pull request by Spinodal**

### CODE REVIEW

File: chapter10/mssw/exercise1.pl [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+use Data::Dumper;
+
+chdir;    # the default is our home directory
+
+#my @sorted = sort { −s $a <=> −s $b } glob '*';
+
+# Whhhhyyyyyy would you make me do thisssss???????!?!?!?!
+my @sorted =
+  map( $_->[0],
+    sort( { $a->[1] <=> $b->[1] } map( [ $_, -s $_ ], glob('*') ) ) );
```
Comment (utopia-2) [inline chapter10/mssw/exercise1.pl:15]: So many parentheses.

File: chapter10/mssw/exercise1.pl [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/perl
+
+use strict;
+use warnings;
+
+use Data::Dumper;
```
Comment (alexdeas) [inline chapter10/mssw/exercise1.pl:6]: You left Data::Dumpers all over the place again.

File: chapter10/mssw/exercise1.pl [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+chdir;    # the default is our home directory
+
+#my @sorted = sort { −s $a <=> −s $b } glob '*';
+
+# Whhhhyyyyyy would you make me do thisssss???????!?!?!?!
```
Comment (alexdeas) [inline chapter10/mssw/exercise1.pl:12]: It's turning you into a perl programmer whether you like it or not.

File: chapter10/mssw/exercise2.out [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Benchmark: running cyborg_friendly_sort, human_friendly_sort for at least 10 CPU seconds...
+cyborg_friendly_sort: 11 wallclock secs ( 5.34 usr +  5.42 sys = 10.76 CPU) @ 56637.83/s (n=609423)
+human_friendly_sort: 11 wallclock secs ( 4.02 usr +  6.57 sys = 10.59 CPU) @ 57469.59/s (n=608603)
```
Comment (alexdeas) [inline chapter10/mssw/exercise2.out:3]: What was your directory like when you globbed it? Potentially if it didn't have very many files the disk operations would have been minimal.

File: chapter10/mssw/exercise2.pl [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+sub human_friendly_sort {
+    my @sorted = sort( { -s $a <=> -s $b } glob('*') );
+}
+
+sub cyborg_friendly_sort {
+    my @sorted =
+      map( $_->[0],
+        sort( { $a->[1] <=> $b->[1] } map( [ $_, -s $_ ], glob('*') ) ) );
+}
```
Comment (alexdeas) [inline chapter10/mssw/exercise2.pl:17]: You went back to the expression form of map, was that so you could add more parentheses?

File: chapter10/mssw/exercise2.out (hunk 4 above)
Comment (utopia-2) [inline chapter10/mssw/exercise2.out:3]: I had a similar result, though evened it out by generating thousands of files. I guess the thing to notice is that the kernel-mode time is higher in the naïve sort. Building a bunch of arrayrefs to cache your results is bound to be more expensive if you have few files or a fast file system.

File: chapter10/mssw/exercise4.pl [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        return \%directory;
+    }
+    warn "$path is neither a file nor a directory\n";
+    return undef;
+}
+
+sub dump_data_for_path {
+    my ( $path, $data, $indent ) = @_;
+    $indent //= 0;
```
Comment (alexdeas) [inline chapter10/mssw/exercise4.pl:31]: Checking this every time seems a bit unnecessary, could maybe just seed the first call with an indent.

File: chapter10/mssw/exercise1.pl (hunk 2 above)
Comment (Spinodal) [inline chapter10/mssw/exercise1.pl:6]: its in my template now :stuck_out_tongue:

File: chapter10/mssw/exercise2.out (hunk 4 above)
Comment (Spinodal) [inline chapter10/mssw/exercise2.out:3]: yea my directory structure was pretty minimal:

```
mssw
  exercise1.pl
  ...
test
 empty
```

File: chapter10/mssw/exercise2.pl (hunk 5 above)
Comment (Spinodal) [inline chapter10/mssw/exercise2.pl:17]: i was just trying to understand what was being done when!
i started by dumping glob, then dumping the mapped glob, then the sorted mapped glob, then ended up with the mapped sorted mapped glob. So i added brackets as i was building it up.

Comment (utopia-2) [inline chapter10/mssw/exercise2.pl:17]: As long as you remember that these operators are right-associative, you get into the mindset of being able to read them right to left.

**Pull request merged**